File: type-theory/2-normalization/lambda/Expression.kt
package lambda

interface Expression {}

data class Variable(val name: String) : Expression {
    override fun toString(): String = name
}

data class Application(val left: Expression, val right: Expression) : Expression {
    override fun toString(): String = "($left $right)"
}

data class Lambda(val variable: Variable, val expression: Expression) : Expression {
    override fun toString(): String = "(\\$variable.$expression)"
}

File: type-theory/1-lambda-expression-parsing-haskell/Lambda/Expression.hs
module Lambda.Expression where

data Expression = Variable    { name        :: String }
                | Application { left        :: Expression, right      :: Expression }
                | Lambda      { variable    :: String,     expression :: Expression }

instance Show Expression where
    show (Variable    str) = str
    show (Application l r) = withParentheses $ show l ++ " " ++ show r
    show (Lambda var expr) = withParentheses $ "\\" ++ var ++ "." ++ show expr

withParentheses :: String -> String
withParentheses value = "(" ++ value ++  ")"
